feat: Adjust common.compat HookLevelLineage collector for new add_extra method

kacpermuda opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->
### TLDR;

Follow-up to #57620 - this PR extends the common.compat provider to support the new Hook Lineage `add_extra` method, ensuring full backward compatibility and consistent hook-level lineage behavior across all Airflow 2.10+ versions.

### Problem

The `add_extra` method for hook lineage collection was introduced in Airflow 3.2 (#57620). However, providers currently support Airflow 2.10+ and may call the new `add_extra` method when running on Airflow versions earlier than 3.2. We previously encountered a similar issue during the transition from `Dataset` to `Asset`, where the hook lineage collector methods changed from `add_input_dataset` to `add_input_asset`. To handle that, we added a compatibility layer in the common.compat provider that manages these naming differences. All other providers were then updated to import the hook lineage collector from common.compat instead of core Airflow, call was changed to the new `_asset` methods and the newer compat provider was made a dependency for those providers.

Without such a compatibility layer, hooks using incompatible code (for example, calling a method that does not exist) would fail on older Airflow versions. This would cause inconsistent behavior and prevent provider hooks from emitting rich lineage metadata. Therefore, I propose applying the same solution here, extending the current compatibility approach.

### Solution

This PR enhances the existing compatibility layer in the common.compat provider to ensure that the hook lineage collector returned from it consistently supports the `add_extra` method across all supported Airflow versions. For Airflow versions that do not natively include `add_extra`, the compatibility layer provides a full implementation of the method, maintaining consistent lineage quality. This is feasible because the hook lineage collector is already operational and collecting lineage data in Airflow 2.10+, so only a small extension was needed to add this capability.

This PR extends current compatibility layer in the `common.compat` provider to ensure that hook lineage collector returned by common.compat provider is equipped with `add_extra` method and behaves consistently across all supported Airflow versions. It implements the full `add_extra` functionality for Airflow versions that don't have it natively to keep consistency in lineage quality. (It's possible because the hook lineage collector is already up and running and is already gathering lineage as the whole structure in airflow core is already there for AF2.10+, so it was just about extending it a bit.)

With this update, provider hooks can safely use:
```
from airflow.providers.common.compat.lineage.hook import get_hook_lineage_collector

get_hook_lineage_collector().add_extra(context, key, value)
```
on any Airflow 2.10+ version, and the behavior will remain identical - provided they depend on the latest common.compat provider version (which should be updated when introducing this call, as per usual practice).

Comprehensive test coverage has been added, covering both standard and edge cases across all Airflow versions included in compatibility testing, ensuring the functionality performs as intended.

This solution allows hook implementations to call `add_extra` without version-specific logic, delivering a consistent lineage experience for users regardless of the Airflow version in use.

### Example Usage
```
from airflow.providers.common.compat.lineage.hook import get_hook_lineage_collector

collector = get_hook_lineage_collector()

# These already worked on all AF 2.10+ versions with previous compat implementation, even when `asset` was still named `dataset`
collector.add_input_asset(context, uri="s3://bucket/input")
collector.add_output_asset(context, uri="s3://bucket/output")

# This will also work on all AF 2.10+ versions now
collector.add_extra(context, "query", "SELECT * FROM table")
collector.add_extra(context, "query_id", "1234")

# And for hook lineage readers:
lineage = collector.collected_assets

# These already worked on all AF 2.10+ versions with previous compat implementation
lineage.inputs - list of AssetLineageInfo
lineage.outputs - list of AssetLineageInfo

# This will also work on all AF 2.10+ versions now
lineage.extra - list of ExtraLineageInfo (now available on all versions!)
```

<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

mobuchowski reviewed (approved):
It's definitely not beautiful... but can't think of better way of making this work on older versions.

ashb reviewed (commented):
I don't think the polyfill approach is one we should use - my first thought is to instead make it a no op only older versions

potiuk reviewed (approved):
LGTM. Waiting for @ashb

ashb commented on providers/common/compat/src/airflow/providers/common/compat/lineage/hook.py:
```diff
@@ -16,24 +16,153 @@
+        collector.__class__,
+        "collected_assets",
+        property(lambda c: _compat_collected_assets(c)),
+    )
```
Does this need to be setattr?

I think this will work just as well
```suggestion
    type(collector).collected_assets =  property(_compat_collected_assets)
```

kacpermuda replied:
Correct, adjusted.

ashb commented on providers/common/compat/src/airflow/providers/common/compat/lineage/hook.py:
```diff
@@ -72,17 +202,34 @@ def collected_assets_compat(collector) -> HookLineage:
-        return get_hook_lineage_collector()
+    # AF 3.0-3.1: needs only polyfill for `add_extra`
+    if not AIRFLOW_V_3_2_PLUS:
+        return _add_extra_polyfill(global_collector)
```
I feel uneasy about this version-based approach, espeically as it isn't the Airflow version we need to be targetting, but the apache-airflow-task-sdk version, as in Airflow core 3.2 timeframe there is a decent chance we will have removed the dependency, meaning TaskSDK could run without apache-airflow-core installed at all.

Elsewhere we have:

```
 if all(
        getattr(collector, asset_method_name, None)
        for asset_method_name in ("add_input_asset", "add_output_asset", "collected_assets", "add_extra")
    ):
```

I think we need to do something similar here

kacpermuda replied:
Agree, thanks for the comment. Added functions that check method/attr presence on collector and we now make decisions based on that. Adjusted the docstring to match that logic. Let me know if that's enough.

ashb reviewed (commented):
Okay with polyfill, but we need to make it duck-typed, not version based (see comment)

kacpermuda committed:
feat: Adjust compat HookLevelLineage for new add_extra method

Extends the current common.compat compatibility layer with support for the new Hook Lineage `add_extra` method, ensuring full backward compatibility and consistent hook-level lineage behavior across all Airflow 2.11+ versions.
